File: packages/wasm/as/assembly/__tests__/context.spec.ts
import { Context } from "../msgpack";


describe("Context class", () => {

  it("pushes and pops values", () => {
    const context: Context = new Context();
    expect(context.length).toStrictEqual(0);
    expect(context.isEmpty()).toStrictEqual(true);
    context.push("property", "string");
    context.push("property", "i32");
    context.push("property", "bool");
    expect(context.length).toStrictEqual(3);
    expect(context.isEmpty()).toStrictEqual(false);
    context.pop();
    context.pop();
    context.pop();
    expect(context.length).toStrictEqual(0);
    expect(context.isEmpty()).toStrictEqual(true);
  });

  it("prints in desired format", () => {
    const context: Context = new Context("Deserializing MyObject");
    context.push("propertyOne", "unknown", "searching for property type");

    expect("\n" + context.toString()).toStrictEqual(
      `
Context: Deserializing MyObject
  at propertyOne: unknown >> searching for property type`
    )

    expect(context.printWithContext("\nInvalid length")).toStrictEqual(
      `
Invalid length
  Context: Deserializing MyObject
    at propertyOne: unknown >> searching for property type`
    )

    context.push("propertyOne", "i32", "type found, reading property");

    expect(context.printWithContext("\nInvalid length")).toStrictEqual(
      `
Invalid length
  Context: Deserializing MyObject
    at propertyOne: i32 >> type found, reading property
      at propertyOne: unknown >> searching for property type`
    )

  });

});

File: packages/wasm/rs/tests/context_spec.rs
use polywrap_wasm_rs::Context;

#[test]
fn push_and_pop_values() {
    let mut context = Context::new();
    assert_eq!(context.get_length(), 0);
    assert!(context.is_empty());
    context.push("property", "string", "");
    context.push("property", "i32", "");
    context.push("property", "bool", "");
    assert_eq!(context.get_length(), 3);
    assert!(!context.is_empty());
    context.pop();
    context.pop();
    context.pop();
    assert_eq!(context.get_length(), 0);
    assert!(context.is_empty());
}

#[test]
fn print_in_desired_format() {
    let mut context = Context::new();
    context.description = "Deserializing MyObject".to_string();
    context.push("propertyOne", "unknown", "searching for property type");

    assert_eq!(context.context_to_string(), "  Context: Deserializing MyObject\n   at propertyOne : unknown >> searching for property type".to_string());

    assert_eq!(context.print_with_context("Invalid length"), "Invalid length\n Context: Deserializing MyObject\n at propertyOne : unknown >> searching for property type");

    context.push("propertyOne", "i32", "type found, reading property");
    assert_eq!(context.print_with_context("\nInvalid length"), "\nInvalid length\n Context: Deserializing MyObject\n at propertyOne : i32 >> type found, reading property\n at propertyOne : unknown >> searching for property type");
}
